compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,101 @@ OMR::Node::isNotCollected()
    return false;
    }
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   if (self()->getOpCode().hasPinningArrayPointer())
+      return false;
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collectable.
+ *
+ * \return
+ *    Return true if the node is collectable, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (!curNode->getType().isAddress())
+      return false;
+
+   while (curNode)
```

Comment:
Remember the "A" in "DAG" is Acyclic :) .

But @andrewcraik and @jdmpapin are right: you can still get an exponential visit if you're descending past commoned nodes (imagine how you would traverse an aternary where both its operands are themselves aternary nodes, etc.). This particular scenario is *probably* pretty rare, but I've hit examples of exponential walks in the past (just not with aternaries). Maybe you only have to memoize the results for aternary nodes (really, any node with more than one address child) to prevent the exponential walk?